### PR TITLE
add configureable oidc subject claim

### PR DESCRIPTION
**Pull request by povilasv**
oidc: allow users to map oidc subject 

### Description

Users can now configure Subject mappings via the clutch config file:
```
      oidc:
        issuer: ...
        client_id: ${CREDENTIALS_OIDC_CLIENT_ID}
        client_secret: ${CREDENTIALS_OIDC_CLIENT_SECRET}
        redirect_url: "${BASE_URL}/v1/authn/callback"
        override_subject_claim_name: email
      session_secret: ${CREDENTIALS_SESSION_SECRET}
```

If these arguments are not set, the old mapper [DefaultClaimsFromOIDCToken](https://pkg.go.dev/github.com/lyft/clutch/backend/service/authn#DefaultClaimsFromOIDCToken) is used.

### Testing Performed

tested locally in a private setup
also added unit tests

### GitHub Issue

Discussed in slack, these flags are similar to Kubernetes API Server's:
```
   - --oidc-groups-claim=groups
   - --oidc-username-claim=email
```

### CODE REVIEW

File: api/config/service/authn/v1/authn.proto [hunk 1]
```diff
@@ -14,6 +14,8 @@ message OIDC {
   string redirect_url = 4 [ (validate.rules).string = {min_bytes : 1} ];
 
   repeated string scopes = 5;
+
+  string override_subject_claim_name= 6;
```
Comment (dschaller) [inline api/config/service/authn/v1/authn.proto:18]: nit: 
```suggestion
  string subject_claim_name_override= 6;
```

File: backend/service/authn/oidc.go [hunk 2]
```diff
@@ -405,22 +405,60 @@ func NewOIDCProvider(ctx context.Context, config *authnv1.Config, tokenStorage S
...
+		return nil, err
+	}
+
+	subjectInt, ok := claims[cc.subjectClaimName]
+	if !ok {
+		return nil, fmt.Errorf("claims did not deserialize with %s field", cc.subjectClaimName)
+	}
+	subject, ok := subjectInt.(string)
+	if !ok {
+		return nil, fmt.Errorf("claims did not deserialize with %s field", cc.subjectClaimName)
```
Comment (dschaller) [inline backend/service/authn/oidc.go:449]: can this error be more around the typing being incorrect since the error is when casting to a string fails?

File: backend/service/authn/oidc.go [hunk 3]
```diff
@@ -405,22 +405,60 @@ func NewOIDCProvider(ctx context.Context, config *authnv1.Config, tokenStorage S
...
+	}
+	if subject == "" {
+		return nil, errors.New("claims did not deserialize with desired fields")
```
Comment (dschaller) [inline backend/service/authn/oidc.go:452]: should this error say something to denote that the subject must be set?

**Comment by povilasv** [comment]
@dschaller Made changes based on your comments, could you take a look again? Thank you :bow:

File: backend/service/authn/oidc.go [hunk 4]
```diff
@@ -405,22 +405,60 @@ func NewOIDCProvider(ctx context.Context, config *authnv1.Config, tokenStorage S
...
+	if !ok {
+		return nil, fmt.Errorf("claims do not contain %s field", cc.subjectClaimName)
+	}
+	subject, ok := subjectInt.(string)
+	if !ok {
+		return nil, fmt.Errorf("claims field %s does not contain string value", cc.subjectClaimName)
+	}
+	if subject == "" {
+		return nil, errors.New("claims field %s is empty")
+	}
```
Comment (dschaller) [inline backend/service/authn/oidc.go:453]: can you add tests for these two error cases as well?

**Comment by dschaller** [comment]
Looks great! Let's add a few more unit tests for the error edge cases just to be sure we don't cause regressions later on and then we can merge :)

File: backend/service/authn/oidc.go (hunk 4 above)
Comment (povilasv) [reply backend/service/authn/oidc.go]: Done :) Caught a bug with them, thank yoU!

File: api/config/service/authn/v1/authn.proto [hunk 5]
```diff
@@ -14,6 +14,8 @@ message OIDC {
   string redirect_url = 4 [ (validate.rules).string = {min_bytes : 1} ];
 
   repeated string scopes = 5;
+
+  string subject_claim_name_override = 6;
```
Comment (danielhochman) [inline api/config/service/authn/v1/authn.proto:18]: can you document the default subject claim when it is not overridden just above this line?

**Review by danielhochman — APPROVED** [review]
lgtm! nicely done, very clean.

**Review by dschaller — APPROVED** [review]

**Comment by povilasv** [comment]
Updated, thanks :)
